doc.go

```diff
@@ -18,121 +18,235 @@
-//     // and can be used to provide an object of type *C
-//     return &C{}
+// Types added to to the container may be consumed by using the Invoke method.
+// Invoke accepts a function which accepts one or more parameters and
+// optionally, returns an error. Dig calls the function with the requested
+// type, instantiating only those types that were requested by the function.
+// The call fails if any type or its dependencies (both direct and transitive)
+// were not available in the container.
+//
+//   err := c.Invoke(func(l *log.Logger) {
+//     // ...
+//   })
+//   if err != nil {
+//     // ...
 //   }
-//   err := c.Provide(constructor)
-//   // dig container is now able to provide *C to any constructor.
-//   //
-//   // However, note that in the current example *A and *B
-//   // have not been provided, meaning that the resolution of type
-//   // *C will result in an error, because types *A and *B can not
-//   // be instantiated.
 //
-// Advanced Provide
+//   err := c.Invoke(func(server *http.Server) error {
+//     // ...
+//   })
+//   if err != nil {
+//     // ...
+//   }
 //
-// // TODO: docs on dig.In usage
-// // TODO: docs on dig.Out usage
+// If the invoked function returns an error, that error is propagated to the
+// caller.
 //
+// Parameter Objects
 //
-// Invoke
+// Constructors declare their dependencies as function parameters. This can
+// very quickly become unreadable if the constructor has a lot of
+// dependencies.
+//
+//   func NewHandler(users *UserGateway, comments *CommentGateway, posts *PostGateway, votes *VoteGateway, authz *AuthZGateway) *Handler {
+//     // ...
+//   }
+//
+// A pattern employed to improve readability in a situation like this is to
+// create a struct which lists all the parameters of the function as fields
+// and changing the function to accept that struct instead. This is referred
+// to as a parameter object.
+//
+// Dig has first class support for parameter objects: any struct that embeds
+// dig.In gets treated as a parameter object. The following is equivalent to
+// the constructor above.
+//
+//   type HandlerParams struct {
+//     dig.In
+//
+//     Users    *UserGateway
+//     Comments *CommentGateway
+//     Posts    *PostGateway
+//     Votes    *VoteGateway
+//     AuthZ    *AuthZGateway
+//   }
+//
+//   func NewHandler(p HandlerParams) *Handler {
+//     // ...
+//   }
 //
-// The Invoke API is the flip side of Provide and used to retrieve types from the container.
+// Handlers can receive any combination of parameter objects and parameters.
 //
-// Invoke looks through the graph and resolves all the constructor parameters for execution.
+//   func NewHandler(p HandlerParams, l *log.Logger) *Handler {
+//     // ...
+//   }
+//
+// Result Objects
+//
+// Result objects are the flip side of parameter objects. These are structs
+// which represent multiple outputs from a single function as fields in the
```

Comment:
As an easy rule to apply in practice, "that" is grammatically proper anywhere you can use it instead of "which" without it sounding wrong.